Launch runs in new browser tabs, break Dagit into Execute + Runs views

bengotow opened the pull request:
This PR refactors the right half of the execution panel a bit to provide a live preview of the execution plan. It's debounced a bit to avoid being re-generated constantly as you type, but should make the editing experience feel more "realtime."

This PR also includes performance improvements to the execution plan view and new tab bar - turns out they re-rendered constantly as you typed and this was less than ideal.

Also note that once you click "Run", the logs appear and from then on, you're looking at the execution plan which was associated with the run, not a preview based on the left-side config. I think this is preferable, but we could also add a way to switch back to the preview?

<img width="1790" alt="image" src="https://user-images.githubusercontent.com/1037212/55032326-df9cdc80-4fcd-11e9-8a4d-5e10cf28dd28.png">

<img width="1790" alt="image" src="https://user-images.githubusercontent.com/1037212/55032313-dc095580-4fcd-11e9-92f2-75233f129dca.png">

bengotow commented:
Ok! This has evolved quite a bit based on our discussions. Here's what's new:

- The execution tab is now just for planning runs and previewing the execution plan. The right sidebar now shows nice error messages if your config is invalid.
- A new "Runs" tab shows a (very very basic) list of your runs. We'll make this look much nicer soon. 
- When you click "Start Execution", Dagit opens a new browser tab and shows the run there.

I also took the time to fix a couple perf issues:

- The execution tab is more performant and it loads the pipeline schema in the background, so there's no awkward white screen when you're switching top-level tabs or solid subsets.
- More React components are "pure" and less of the UI refreshes as you type in the config editor. It was actually a bit janky before.

On the python side there's just one change:
- pipelineRun is now pipelineRunOrError and there's a new PipelineRunNotFoundError. The current impl was actually broken so I think this change won't affect anyone.


TODO:
 - How do you view config for a run? We store it, but it's no longer displayed anywhere.

<img width="1355" alt="image" src="https://user-images.githubusercontent.com/1037212/55120709-514f5600-50b4-11e9-95b8-df4210fd1a66.png">
<img width="1355" alt="image" src="https://user-images.githubusercontent.com/1037212/55120838-c6bb2680-50b4-11e9-9e4b-0c509daf55b4.png">

mgasner commented:
Resolves #725

mgasner commented:
It might be nice to have a config pop out in one of the columns in the eventual run view.

schrockn commented:
Oh and please update CHANGES.rst for this change

bengotow commented on js_modules/dagit/src/execute/ExecutionStartButton.tsx:
```diff
@@ -89,32 +90,43 @@ export default class ExecutionStartButton extends React.Component<
 }
 
-const Wrapper = styled.div<{ disabled?: boolean }>`
+const Wrapper = styled.div<{ state: "ready" | "disabled" | "running" }>`
```
Mmm we could potentially re-use the pipeline state enum here, but the execute button state is a bit more transient since it becomes "ready" again as soon as the pipeline is kicked off, and "disabled" means the dagit process is gone. I'll convert it into an enum in this file though.

bengotow commented on js_modules/dagit/src/configeditor/ConfigEditor.tsx:
```diff
@@ -23,7 +23,7 @@ import { LintJson as YamlModeLintJson } from "./codemirror-yaml/mode";
 
 interface IConfigEditorProps {
-  pipeline: ConfigEditorPipelineFragment;
+  pipeline: ConfigEditorPipelineFragment | "loading";
```
yeah I wasn't sure what we should do here. These components can be displayed without a `pipeline` (which is preferable to not displaying them at all until the pipeline has been fetched) but don't become interactable / fully usable until the pipeline is loaded. I didn't want to make the prop optional since it might seem like using the component without a pipeline was a valid use case. I figured `pipeline | "loading"` made the two available states more clear than `pipeline | undefined`.

I'll switch 'em to be optional - I think that is more standard.

bengotow commented:
OK addressed the following:

- Added line to CHANGES

- When you click to expand a full log message in a model, is you press escape this actually navigates to the overall runs page, rather than the individual run page. This is not a good user experience.
(This one was actually the pipeline dropdown in the navbar thinking you were clearing it's selection. wild.)

- We can get into a state, especially in the airline demo, where we actually instigate either pipeline construction or initialization which is expense. So what happens is that it takes a while for the config to verify that it is valid or not. If you press the start execution button in the interim, it get into a bad state where execution has initiated but dagit doesn't know it. This may be worth a brief chat to demonstrate as the interaction required to trigger and then resolve it is a little intricate.

No this definitely makes sense - I moved the query that fetches the pipeline metadata into the background so that the UI doesn't flicker when you switch to the Execute tab. I'll add proper handling of the intermediate state.

I think these will take a bit more work - I can do them on this PR or another one but probably not till Monday or Tues! I think we'll want to redo the log UI a bit and add a top bar with the run ID, overall execution time, date, etc.

    It would be fantastic to instigate a rerun of an execution with identical config easily from an existing run page.
    More information on the Runs page. Time and pipeline name would be a big improvement.
    More columns in the log view representing the structured information we stream to client, in addition to the raw log message.
